### DISCUSSION TITLE
How was the rendering on WinForms improved?

### DISCUSSION THREAD

**Discussion by proneon267** [beeware/toga#2107] [Q&A] [answered]
Not too long ago, toga apps on WinForms looked quite blurry and other issues like incorrect size font rendering were present. 

These issues were especially visible in Hi-DPI displays. My first guess is that these issues were caused by windows scaling. 

So, how were these issues resolved?

**Comment by mhsmith** [comment, marked as answer]
We've been systematically adding tests on all widgets and fixing the issues they expose, including scaling issues. Here's where the scaling code currently exists:

https://github.com/beeware/toga/blob/237ccee6454351f36a4ef91f6b5ca82395e84819/winforms/src/toga_winforms/widgets/base.py#L15

**Reply by proneon267** [reply]
Thank you for helping. I have some questions regarding the DPI scaling:
https://github.com/beeware/toga/blob/237ccee6454351f36a4ef91f6b5ca82395e84819/winforms/src/toga_winforms/widgets/base.py#L19

Here, 96 PPI is used because it is the standard Windows DPI value? or because the app is in DPI Unaware mode(https://learn.microsoft.com/en-us/windows/win32/hidpi/high-dpi-desktop-application-development-on-windows?redirectedfrom=MSDN#dpi-unaware)?

Also, I had thought that the maximum of the two (DpiX and DpiY) is used to calculate the scale factor to account for monitors with different horizontal and vertical DPI settings. Is this not true?

**Reply by mhsmith** [reply]
Toga runs in DPI-aware mode:

https://github.com/beeware/toga/blob/237ccee6454351f36a4ef91f6b5ca82395e84819/winforms/src/toga_winforms/app.py#L64-L65

So when the display DPI is different from the standard value of 96, Toga has to adjust all sizes and locations accordingly.

> Also, I had thought that the maximum of the two (DpiX and DpiY) is used to calculate the scale factor to account for monitors with different horizontal and vertical DPI settings.

I don't think anyone's used Windows on a monitor like that since about 1991.

**Reply by mhsmith** [reply]
Since PR #2020, all DPI corrections are done internally by each backend. The Toga core, and its public API, works entirely in virtual [CSS pixels](https://hacks.mozilla.org/2013/09/css-length-explained/).

**Reply by proneon267** [reply]
Thanks for helping out :)
